pkg/controller/master/node/node_down_controller.go

```diff
@@ -3,278 +3,343 @@ package node
+	var cond *corev1.NodeCondition
+	for i := range conditions {
+		c := conditions[i]
+		if c.Type == conditionType {
+			cond = &c
+			break
+		}
 	}
+	return cond
+}
 
-	if !vmForceResetPolicy.Enable {
-		return node, nil
+// We treat the node not ready (False/Unknown) as node down
+// In this situation, we will wait for the extra timeout and then add taints to trigger migration and cleanup
+func (h *nodeDownHandler) checkNodeReady(node *corev1.Node) error {
+	// get Ready condition
+	cond := getNodeCondition(node.Status.Conditions, corev1.NodeReady)
+	if cond == nil {
+		return fmt.Errorf("can't find %s condition in node %s", corev1.NodeReady, node.Name)
 	}
 
-	// if we haven't waited for vmForceResetPolicy.Period seconds, we enqueue event again
-	if time.Since(cond.LastTransitionTime.Time) < time.Duration(vmForceResetPolicy.Period)*time.Second {
-		deadline := cond.LastTransitionTime.Add(time.Duration(vmForceResetPolicy.Period) * time.Second)
-		logrus.Debugf("Enqueue node event again at %v", deadline)
-		h.nodes.EnqueueAfter(node.Name, time.Until(deadline))
-		return node, nil
-	}
+	switch cond.Status {
+	case corev1.ConditionFalse:
+	case corev1.ConditionUnknown:
+		// wait for extra timeout to
+		vmForceResetPolicy, err := h.fetchVMForceResetPolicy()
+		if err != nil {
+			return fmt.Errorf("failed to fetch VMForceResetPolicy setting: %w", err)
+		}
 
-	// get VMI pods on unhealthy node
-	pods, err := h.pods.List(corev1.NamespaceAll, metav1.ListOptions{
-		LabelSelector: labels.Set{
-			kubevirtv1.AppLabel: "virt-launcher",
-		}.String(),
-		FieldSelector: "spec.nodeName=" + node.Name,
-	})
-	if err != nil {
-		return node, err
-	}
+		if !vmForceResetPolicy.Enable {
+			// skip if the setting is disabled
+			return nil
+		}
 
-	var errs error
-	pvcSet := set.New[string]()
-	gracePeriod := int64(0)
-	for i := range pods.Items {
-		pod := pods.Items[i]
-		logrus.Debugf("force delete pod %s/%s and VolumeAttachments", pod.Namespace, pod.Name)
-
-		if err := h.pods.Delete(
-			pod.Namespace,
-			pod.Name,
-			&metav1.DeleteOptions{
-				GracePeriodSeconds: &gracePeriod,
-			}); err != nil {
-			errs = multierr.Append(errs, fmt.Errorf("failed to delete pod %s/%s: %w", pod.Namespace, pod.Name, err))
-			continue
+		if getNodeTaint(node.Spec.Taints, virtconfig.NodeDrainTaintDefaultKey) != nil && getNodeTaint(node.Spec.Taints, corev1.TaintNodeOutOfService) != nil {
+			// already added both taints, no need to process again
+			return nil
 		}
 
-		names, err := h.getPVCsName(&pod)
+		// try to add Kubevirt drain taint after waiting extra timeout
+		if waitEnough, err := h.addKubevirtTaintAfterExtraTimeout(node, cond, vmForceResetPolicy.Period); err != nil {
+			return err
+		} else if !waitEnough {
+			return nil
+		}
+
+		node, err = h.createOrUpdateNodeCondition(node, HarvesterNodeCondDrained, corev1.ConditionFalse, HarvesterNodeDrainedCondReasonDraining, HarvesterNodeDrainedCondMsg)
```

Comment:
Will this cause busy loop when node down?  There is no reflect.deepEqual check in `createOrUpdateNodeCondition`, when we invoke `createOrUpdateNodeCondition` it will trigger another reconcile loop in this controller, right?